### ISSUE TITLE
(chakra) recursive evaluation

### ISSUE THREAD

**Issue by ViceIce** [Taritsyn/JavaScriptEngineSwitcher#72]
I'm trying to a simple module loading feature which requires recursive script execution:

`net -> js -> net -> js`

This is currently not possible, because the scriptdispatcher is blocked by the first `eval`.
https://github.com/Taritsyn/JavaScriptEngineSwitcher/blob/753deb75fa62eb82ef1f5ea019a98df708c4eb7d/src/JavaScriptEngineSwitcher.ChakraCore/ScriptDispatcher.cs#L152

My current workaround is to use eval, which i don't like. Is there any way to make recursive calls work?

**Comment by Taritsyn** [comment]
Hello, Michael!

You could give an example of JS code that causes this error.

**Comment by ViceIce** [comment]
Simple test case:
```cs
_engine.EmbedHostObject("_test", (Action<string>)(c => _engine.Evaluate(c));
_engine.Evaluate("_test('var test;')");
```

**Comment by Taritsyn** [comment]
Why are you doing this? You can use a `eval` keyword.

**Comment by ViceIce** [comment]
What about this:

```cs
_engine.EmbedHostObject("_test", (Action<string>)(c => _engine.ExecuteFile($"dist/{c}.js"));
_engine.Evaluate("_test('test')");
```

I'm trying to implement a simple module system.

**Comment by Taritsyn** [comment]
Try this variant:

```csharp
Func<string, string> readFileContent = File.ReadAllText;

_engine.EmbedHostObject("readFileContent", readFileContent);
_engine.Execute(@"function loadModule(moduleName) {
    var modulePath = 'dist/' + moduleName + '.js';
    var moduleContent = readFileContent(modulePath);

    return new Function(moduleContent)();
}");
_engine.CallFunction("loadModule", "test");
```

**Comment by ViceIce** [comment]
That's my workaround 😄

**Comment by Taritsyn** [comment]
Only here, instead of the `eval(…)` expression used the `new Function(…)()`, which is more safe.

**Comment by Taritsyn** [comment]
When there is free time, I will deal with this problem.

**Comment by ViceIce** [comment]
**Solution**
https://github.com/Taritsyn/JavaScriptEngineSwitcher/blob/753deb75fa62eb82ef1f5ea019a98df708c4eb7d/src/JavaScriptEngineSwitcher.ChakraCore/ScriptDispatcher.cs#L147-L153

1. Inline the execution if we don't specify a max stack size.
2. Check, if we are the dispatcher thread and inline the execution.

Which solution do you prefer, or another solution? I can send a PR.

**Comment by Taritsyn** [comment]
> Which solution do you prefer, or another solution? I can send a PR.

I'll deal with the problem myself, but later.

**Comment by ViceIce** [comment]
OK, the `new Function` solution is a little bit annoying, because i don't have a filename in the stack traces.

So it is very difficult to find the right source file.

**Comment by Taritsyn** [comment]
Hello, Michael!

This error is fixed in [version 3.0.8](https://github.com/Taritsyn/JavaScriptEngineSwitcher/releases/tag/v3.0.8).
